### PR TITLE
Master/Slave radius proxy and degraded workflow

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Proxy radius request when the server is in slave mode.
It include a degraded radius configuration in order to reply even if the main radius server is down.

When the main server is up, the slave radius server proxy to the main one and store the reply in the radreply table.

If it 802.1x the nthash key is shared all accross the servers member and in degraded situation radius use this hash and reply with the latest attributes coming from previous authentication (radreply table).

If it's mac-auth we reply with the previous attributes.

# Impacts
Master Slave configuration.

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
## New Features
* Master/Slave radius proxy and degraded workflow

### CODE REVIEW

**Comment by jrouzierinverse** [comment]
Can you rebase?
Also, there is a circular dependency in ```pf::cluster``` causing the perl tests to break.

**Comment by extrafu** [comment]
@fdurand Please rebase so we can merge it

**Comment by extrafu** [comment]
@fdurand Please rebase.

**Comment by extrafu** [comment]
@jrouzierinverse Please review.

File: db/pf-schema-X.Y.Z.sql [hunk 1]
```diff
@@ -690,6 +690,20 @@ CREATE TABLE radacct_log (
...
+
+CREATE TABLE radreply (
+  id int(11) unsigned NOT NULL auto_increment,
+  tenant_id int NOT NULL DEFAULT 1,
+  username varchar(64) NOT NULL default '',
+  attribute varchar(64) NOT NULL default '',
+  op char(2) NOT NULL DEFAULT '=',
+  value varchar(253) NOT NULL default '',
+  PRIMARY KEY (`tenant_id`, `username`)
+);
```
Comment (jrouzierinverse) [inline db/pf-schema-X.Y.Z.sql:704]: Go the following error.
"ERROR 1075 (42000) at line 695: Incorrect table definition; there can be only one auto column and it must be defined as a key"

File: lib/pf/cluster.pm [hunk 2]
```diff
@@ -790,6 +790,14 @@ sub getDBMaster {
...
 }
 
+sub getDBMaster {
+    if (defined(${pf::config::cluster::getClusterConfig(${pf::config::cluster::getClusterConfig($clusters_hostname_map{$host_id})}{CLUSTER}{masterdb})}{CLUSTER}{management_ip})) {
+        return ${pf::config::cluster::getClusterConfig(${pf::config::cluster::getClusterConfig($clusters_hostname_map{$host_id})}{CLUSTER}{masterdb})}{CLUSTER}{management_ip};
+    } else {
+        return $FALSE;
+    }
+}
+
```
Comment (jrouzierinverse) [inline lib/pf/cluster.pm]: This is function getDBMaster is duplicated.

File: lib/pf/services/manager/radiusd_child.pm [hunk 3]
```diff
@@ -1233,46 +1277,57 @@ Generate the configuration for eap choice
...
 EOT
+    }
+    if ($suffix ne "") {
+    $$authentication_auth_type .= <<"EOT";
+        Auth-Type $suffix {
+            $suffix
         }
-
+EOT
+    }
```
Comment (jrouzierinverse) [inline lib/pf/services/manager/radiusd_child.pm:1330]: This file fails coding-style.t

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]

**Comment by fdurand** [comment]
Tested and fixed. Ready for another review.

**Comment by extrafu** [comment]
@jrouzierinverse Review again and merge if all is fine.

**Pull request merged**